Rubberduck.VBEEditor/SafeComWrappers/VB/Abstract/IHostApplication.cs

```diff
@@ -1,36 +1,92 @@
+        /// Gets the name of the application.
         /// </summary>
-        /// <param name="declaration">The Declaration object for the method to be executed.</param>
-        [Obsolete("Use ExecuteCode in TypeLibAPI instead", true)]
-        void Run(dynamic declaration);
+        /// <remarks>
+        /// This is needed only to circumvent the problem that HostApplicationBase-derived types
+        /// cannot be used outside assembly boundaries because the type is generic.
+        /// </remarks>
+        string ApplicationName { get; }
 
         /// <summary>
-        /// Executes a VBA function by name, with specified parameters, and returns a result.
+        /// Gets data for the host-specific documents not otherwise exposed via VBIDE API
         /// </summary>
-        /// <param name="name"></param>
-        /// <param name="args"></param>
-        /// <returns></returns>
         /// <remarks>
-        /// May not be available in all host applications.
+        /// Not all properties are available via VBIDE API. While some properties may be
+        /// accessed via the <see cref="IVBComponent.Properties"/>, there are problems
+        /// with using those properties when the document is not in a design mode. For
+        /// that reason, it's better to get the data using host's object model instead.
         /// </remarks>
-        [Obsolete("Use ExecuteCode in TypeLibAPI instead", true)]
-        object Run(string name, params object[] args);
+        IEnumerable<IHostDocument> GetDocuments();
 
         /// <summary>
-        /// Gets the name of the application.
+        /// Gets data for a host-specific document not otherwise exposed via VBIDE API
         /// </summary>
-        /// <remarks>
-        /// This is needed only to circumvent the problem that HostApplicationBase-derived types
-        /// cannot be used outside assembly boundaries because the type is generic.
-        /// </remarks>
-        string ApplicationName { get; }
+        /// <param name="moduleName"><see cref="QualifiedModuleName"/> representing a VBComponent object</param>
+        /// <returns><see cref="IHostDocument"/> data</returns>
+        /// <inheritdoc cref="GetDocuments"/>
+        IHostDocument GetDocument(QualifiedModuleName moduleName);
+    }
+
+    public enum DocumentState
+    {
+        /// <summary>
+        /// The document is not open and its accompanying <see cref="IVBComponent"/> may not be available.
+        /// </summary>
+        Closed,
+        /// <summary>
+        /// The document is open in design mode.
+        /// </summary>
+        DesignView,
+        /// <summary>
+        /// The document is open in non-design mode. It may not be safe to parse the document in this state.
+        /// </summary>
+        ActiveView
+    }
+
+    public interface IHostDocument : IDisposable
+    {
+        string Name { get; }
+        string ClassName { get; }
+        WeakReference<object> Target { get; }
```

Comment:
Thinking about it - I think this is not ideal. This should be probably `bool TryGetTarget(out object)` instead. This way no references is held  within the implementation of `IHostDocument` and the consumer can dispose of the object. 

OTOH, YAGNI...